README
This is a version of bacnet_server to access the remote modbus server at 140.159.153.159
It has been renamed EES4100_Bridge

The program was modified to have the threads move input data from the Modbus side to the Bacnet side.
Tested and working.

Additional cleaning up of indents and comments


NEWS
EES4100_Bridge is working.


src/modbus_client.c
/*Modified modbus client sample from libmodbus.org*/
/*added error checking to basic sample with messaging
* changed context variable to ctx 
* port number is 502 (default) and loop-back address changed to server address
* user must be SU to access that part of memory
*/

#include <stdio.h>
#include <modbus-tcp.h>
#include <errno.h>

int main(void) {

	modbus_t *ctx;
	uint16_t tab_reg[32];

	ctx = modbus_new_tcp("140.159.153.159", 502);  //port requires admin privileges
	//error checking
	if (ctx == NULL)  {
		fprintf(stderr, "Unable to allocate libmodbus context\n");
		return -1;
	}

	modbus_connect(ctx);
	//error checking
        if (modbus_connect(ctx) == -1) {
		fprintf(stderr, "Connection failed: %s\n", modbus_strerror(errno));
		modbus_free(ctx);
		return -1;
	
	}

	else {
		printf("Connected to server\n");
	}

/* Read 2 registers from the address 44 and 45 */
	modbus_read_registers(ctx, 44, 2, tab_reg);

	printf("Register value:%04x\n", tab_reg[0]);

	modbus_close(ctx);
	modbus_free(ctx);
}


src/EES4100_Bridge.c
/* EES4100 Bridge Application with patch for link lists applied
*  Using remote server addresses
*  comments:
	Changed some names for clarity
*	Added dereferencing for the BACnet analog input read operations
*  	Cleaned up the presentation*/

#include <stdio.h>

#include <libbacnet/address.h>	/*bacnet headers */
#include <libbacnet/device.h>
#include <libbacnet/handlers.h>
#include <libbacnet/datalink.h>
#include <libbacnet/bvlc.h>
#include <libbacnet/client.h>
#include <libbacnet/txbuf.h>
#include <libbacnet/tsm.h>
#include <libbacnet/ai.h>
#include "bacnet_namespace.h"

#include "modbus-tcp.h"		/*Modbus header includes all else required */

#define SERVER_PORT		502	/* Modbus default port number */
//#define SERVER_ADDRESS        "127.0.0.1"     /* loopback address for testing*/
#define SERVER_ADDRESS	"140.159.153.159"	/* remote Modbus server address */

#define BACNET_DEVICE_ID	44	/* assigned device Id */
#define INST_NUM		2	/* total Analog Instances */
//#define BACNET_INSTANCE_NO    12      /* used for self-testing*/
#define BACNET_PORT		0xBAC1	/* default Bacnet UDP port */
#define BACNET_INTERFACE	"lo"	/* loopback mode */
#define BACNET_DATALINK_TYPE	"bvlc"	/* Bacnet virtual link control */
#define BACNET_SELECT_TIMEOUT_MS    1	/* msec */
#define RUN_AS_BBMD_CLIENT	1	/* for if statement below */

#if RUN_AS_BBMD_CLIENT		/* BBMD =>talking to foreign device */
#define BACNET_BBMD_PORT	0xBAC0
#define BACNET_BBMD_ADDRESS	"140.159.160.7"	/* when BBMD operating */
#define BACNET_BBMD_TTL		90
#endif

//--------Setting up threads: Modbus client and BACnet server-----

/*1. Define linked list object type*/

typedef struct s_AI_object AI_object;
struct s_AI_object {
    uint16_t number;
    AI_object *next;		//next points to address for analog data
};

static AI_object *list_heads[INST_NUM];	//there will be 2 instances

/*2. Use of list lock to share the list between threads*/

static pthread_mutex_t timer_lock = PTHREAD_MUTEX_INITIALIZER;
static pthread_mutex_t list_lock = PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t list_data_ready = PTHREAD_COND_INITIALIZER;
static pthread_cond_t list_data_flush = PTHREAD_COND_INITIALIZER;

/*3. Modbus server side: Add objects to list- make sure it gets locked/unlocked*/

static void add_to_list(AI_object ** list_heads, uint16_t number)
{
    pthread_mutex_lock(&list_lock);	//LOCK to begin list operation
    printf("adding to list\n");		//test print 

    AI_object *last_object, *temp_object;	/* 2nd pointer variable for array */
    temp_object = malloc(sizeof(AI_object));	/* allocate of memory for data */
    temp_object->number = number;	/* number dereferenced to save its value */
    temp_object->next = NULL;		/* next pointer set to NULL */

    if (*list_heads == NULL) {		/* check list head addr value */
	*list_heads = temp_object;	/* put temp object at the head if empty */
    } else {
	last_object = *list_heads;	/* if not empty go to next data */
	while (last_object->next) {
	    last_object = last_object->next;	/* move through list */
	}
	last_object->next = temp_object;	/* new data */
    }

    pthread_mutex_unlock(&list_lock);	/* unlock list again */
    pthread_cond_signal(&list_data_ready);	/* for list flush operation only */
}

/*4. BACnet Client side: Take data from the top of the list*/

static AI_object *list_get_first(AI_object ** list_heads)
{
    AI_object *first_object;
    first_object = *list_heads;
    *list_heads = (*list_heads)->next;	/* get next value */
    return first_object;
}

/*5. BACnet READ PROPERTY */

static int Update_Analog_Input_Read_Property(BACNET_READ_PROPERTY_DATA
					     * rpdata)
{
    AI_object *first_object;	/* define object */
    int instance =
	bacnet_Analog_Input_Instance_To_Index(rpdata->object_instance);
    if (rpdata->object_property != bacnet_PROP_PRESENT_VALUE) {
	goto not_pv;		/* check for value */
    }
    pthread_mutex_lock(&list_lock);

    if (list_heads[instance] == NULL) {	/*and check the instance as well */
	pthread_mutex_unlock(&list_lock);
	goto not_pv;
    }
    first_object = list_get_first(&list_heads[instance]);
    printf("AI_Present_Value request for instance %i\n", instance);

    	/* need  a current value object to push through to client */
    bacnet_Analog_Input_Present_Value_Set(instance, first_object->number);
    free(first_object);		/* get ready for next data */
    pthread_mutex_unlock(&list_lock);

  not_pv:			/*If not present value or instance null, skips here */
    printf("Not Present_Value\n");	//test print
    return bacnet_Analog_Input_Read_Property(rpdata);
}

/* SERVER OBJECTS*/

static bacnet_object_functions_t server_objects[] = {
    {bacnet_OBJECT_DEVICE,
     NULL,
     bacnet_Device_Count,
     bacnet_Device_Index_To_Instance,
     bacnet_Device_Valid_Object_Instance_Number,
     bacnet_Device_Object_Name,
     bacnet_Device_Read_Property_Local,
     bacnet_Device_Write_Property_Local,
     bacnet_Device_Property_Lists,
     bacnet_DeviceGetRRInfo,
     NULL,	/* Iterator */
     NULL,	/* Value_Lists */
     NULL,	/* COV */
     NULL,	/* COV Clear */
     NULL 	/* Intrinsic Reporting */ },
    {bacnet_OBJECT_ANALOG_INPUT,
     bacnet_Analog_Input_Init,
     bacnet_Analog_Input_Count,
     bacnet_Analog_Input_Index_To_Instance,
     bacnet_Analog_Input_Valid_Instance,
     bacnet_Analog_Input_Object_Name,
     Update_Analog_Input_Read_Property,
     bacnet_Analog_Input_Write_Property,
     bacnet_Analog_Input_Property_Lists,
     NULL 	/* ReadRangeInfo */ ,
     NULL 	/* Iterator */ ,
     bacnet_Analog_Input_Encode_Value_List,
     bacnet_Analog_Input_Change_Of_Value,
     bacnet_Analog_Input_Change_Of_Value_Clear,
     bacnet_Analog_Input_Intrinsic_Reporting},
    {MAX_BACNET_OBJECT_TYPE}
};

/*REGISTER AS A FOREIGN DEVICE*/

static void register_with_bbmd(void)
{
#if RUN_AS_BBMD_CLIENT
		/* Thread safety: Shares data with datalink_send_pdu */
    bacnet_bvlc_register_with_bbmd(bacnet_bip_getaddrbyname
				   (BACNET_BBMD_ADDRESS),
				   htons(BACNET_BBMD_PORT),
				   BACNET_BBMD_TTL);
#endif
}

/*SETTING UP MINUTE AND SECOND TIMERS*/

static void *minute_tick(void *arg)
{
    while (1) {
	pthread_mutex_lock(&timer_lock);

		/* Expire addresses once the TTL has expired */
	bacnet_address_cache_timer(60);

		/* Re-register with BBMD once BBMD TTL has expired */
	register_with_bbmd();

		/*Update addresses for notification class recipient list 
		 * Requred for INTRINSIC_REPORTING
		 * bacnet_Notification_Class_find_recipient(); */

		/* Sleep for 1 minute */
	pthread_mutex_unlock(&timer_lock);
	sleep(60);
    }
    return arg;
}
static void *second_tick(void *arg)
{
    while (1) {
	pthread_mutex_lock(&timer_lock);
		/* Invalidates stale BBMD foreign device table entries */
	bacnet_bvlc_maintenance_timer(1);
		/* Transaction state machine: Responsible for retransmissions 
	 	 * and ack checking for confirmed services */
	bacnet_tsm_timer_milliseconds(1000);
		/*removed more bacnet functions not used here */
		/* Sleep for 1 second */
	pthread_mutex_unlock(&timer_lock);
	sleep(1);
    }
    return arg;
}

/*START OF MODBUS THREAD*/

static void *modbus_side(void *arg)	/*allocate and initialize a structure*/
{
    uint16_t tab_reg[32];
    int rv;			/* return value*/
    int i;			/* iteration */
    modbus_t *ctx;		/* pointer to structure*/

    printf("got to the modbus thread\n");	//test print       

  modbus_restart:		/* retry connection at this point*/
				/* establish the context for communication*/
    ctx = modbus_new_tcp(SERVER_ADDRESS, SERVER_PORT);

    if (ctx == NULL) {		/* structure not allocated*/
	fprintf(stderr, "Unsuccessful allocation and initialization\n");
	sleep(1);
	goto modbus_restart;
    }
    if (modbus_connect(ctx) == -1) {	/* establish connection status*/
	fprintf(stderr, "Unsuccessful connection to server: %s\n",
		modbus_strerror(errno));
	modbus_free(ctx);
	modbus_close(ctx);
	sleep(1);
	goto modbus_restart;
    } else {
	fprintf(stderr, "Successful connection to server\n");
    }

/* READ MODBUS REGISTERS*/

    while (1) {			/* assigned address info*/
	rv = modbus_read_registers(ctx, BACNET_DEVICE_ID, INST_NUM,
				   tab_reg);
	printf("got into the read registers: rv value is: %d\n", rv);//test
	if (rv == -1) {
	    fprintf(stderr, "Register read failed: %s\n",
		    modbus_strerror(errno));
	    modbus_free(ctx);
	    modbus_close(ctx);
	    sleep(1);
	    goto modbus_restart;
	}
	for (i = 0; i < rv; i++) {
	    add_to_list(&list_heads[i], tab_reg[i]);
	    printf("Register[%d] = [%d] (0x%X)\n", i, tab_reg[i],
		   tab_reg[i]);
	}
	usleep(100000);		/* 100ms sleep */
	//return 0;  designed to loop forever
    }
    printf("got to the end of modbus thread \n");	//test print       
}	//should never see end of modbus thread

/* BACnet APDU SERVICE */

#define BN_UNC(service, handler) \
    bacnet_apdu_set_unconfirmed_handler(		\
		    SERVICE_UNCONFIRMED_##service,	\
		    bacnet_handler_##handler)
#define BN_CON(service, handler) \
    bacnet_apdu_set_confirmed_handler(			\
		    SERVICE_CONFIRMED_##service,	\
		    bacnet_handler_##handler)

/* MAIN */

int main(int argc, char **argv)
{
    uint8_t rx_buf[bacnet_MAX_MPDU];
    uint16_t pdu_len;
    BACNET_ADDRESS src;
    pthread_t minute_tick_id, second_tick_id;
    pthread_t modbus_start_id;		/*added modbus thread */

    bacnet_Device_Set_Object_Instance_Number(BACNET_DEVICE_ID);
    bacnet_address_init();

    printf("got to the main\n");	//test print

    bacnet_Device_Init(server_objects);	/* Setup device objects */
    BN_UNC(WHO_IS, who_is);
    BN_CON(READ_PROPERTY, read_property);

    bacnet_BIP_Debug = true;
    bacnet_bip_set_port(htons(BACNET_PORT));
    bacnet_datalink_set(BACNET_DATALINK_TYPE);
    bacnet_datalink_init(BACNET_INTERFACE);
    atexit(bacnet_datalink_cleanup);
    memset(&src, 0, sizeof(src));

    register_with_bbmd();

    bacnet_Send_I_Am(bacnet_Handler_Transmit_Buffer);

    pthread_create(&minute_tick_id, 0, minute_tick, NULL);
    pthread_create(&second_tick_id, 0, second_tick, NULL);
    pthread_create(&modbus_start_id, 0, modbus_side, NULL);

/*holding the lock to ensure synchronised operations */

    while (1) {
	pdu_len = bacnet_datalink_receive(&src, rx_buf, bacnet_MAX_MPDU,
					  BACNET_SELECT_TIMEOUT_MS);
	if (pdu_len) {
	    /* May call any registered handler.
	     * Thread safety: May block, however we still need to guarantee
	     * atomicity with the timers, so hold the lock anyway */
	    pthread_mutex_lock(&timer_lock);
	    bacnet_npdu_handler(&src, rx_buf, pdu_len);
	    pthread_mutex_unlock(&timer_lock);
	}
    }
    return 0;
}				//END OF PROGRAM


src/bacnet_namespace.h
#define bacnet_address_bind_request address_bind_request
#define bacnet_address_cache_timer address_cache_timer
#define bacnet_address_init address_init
#define bacnet_address_match address_match
#define bacnet_Analog_Input_Change_Of_Value Analog_Input_Change_Of_Value
#define bacnet_Analog_Input_Change_Of_Value_Clear Analog_Input_Change_Of_Value_Clear
#define bacnet_Analog_Input_Count Analog_Input_Count
#define bacnet_Analog_Input_Encode_Value_List Analog_Input_Encode_Value_List
#define bacnet_Analog_Input_Index_To_Instance Analog_Input_Index_To_Instance
#define bacnet_Analog_Input_Init Analog_Input_Init
#define bacnet_Analog_Input_Instance_To_Index Analog_Input_Instance_To_Index
#define bacnet_Analog_Input_Intrinsic_Reporting Analog_Input_Intrinsic_Reporting
#define bacnet_Analog_Input_Object_Name Analog_Input_Object_Name
#define bacnet_Analog_Input_Present_Value_Set Analog_Input_Present_Value_Set
#define bacnet_Analog_Input_Property_Lists Analog_Input_Property_Lists
#define bacnet_Analog_Input_Read_Property Analog_Input_Read_Property
#define bacnet_Analog_Input_Valid_Instance Analog_Input_Valid_Instance
#define bacnet_Analog_Input_Write_Property Analog_Input_Write_Property
#define bacnet_apdu_set_abort_handler apdu_set_abort_handler
#define bacnet_apdu_set_confirmed_ack_handler apdu_set_confirmed_ack_handler
#define bacnet_apdu_set_confirmed_handler apdu_set_confirmed_handler
#define bacnet_apdu_set_error_handler apdu_set_error_handler
#define bacnet_apdu_set_reject_handler apdu_set_reject_handler
#define bacnet_apdu_set_unconfirmed_handler apdu_set_unconfirmed_handler
#define bacnet_BIP_Debug BIP_Debug
#define bacnet_bip_getaddrbyname bip_getaddrbyname
#define bacnet_bip_set_port bip_set_port
#define bacnet_bvlc_maintenance_timer bvlc_maintenance_timer
#define bacnet_bvlc_register_with_bbmd bvlc_register_with_bbmd
#define bacnet_datalink_cleanup datalink_cleanup
#define bacnet_datalink_init datalink_init
#define bacnet_datalink_receive datalink_receive
#define bacnet_datalink_set datalink_set
#define bacnet_dcc_timer_seconds dcc_timer_seconds
#define bacnet_Device_Count Device_Count
#define bacnet_DeviceGetRRInfo DeviceGetRRInfo
#define bacnet_Device_Index_To_Instance Device_Index_To_Instance
#define bacnet_Device_Init Device_Init
#define bacnet_Device_local_reporting Device_local_reporting
#define bacnet_Device_Object_Name Device_Object_Name
#define bacnet_Device_Property_Lists Device_Property_Lists
#define bacnet_Device_Read_Property_Local Device_Read_Property_Local
#define bacnet_Device_Set_Object_Instance_Number Device_Set_Object_Instance_Number
#define bacnet_Device_Valid_Object_Instance_Number Device_Valid_Object_Instance_Number
#define bacnet_Device_Write_Property_Local Device_Write_Property_Local
#define bacnet_dlenv_maintenance_timer dlenv_maintenance_timer
#define bacnet_handler_cov_task handler_cov_task
#define bacnet_handler_cov_timer_seconds handler_cov_timer_seconds
#define bacnet_handler_i_am_bind handler_i_am_bind
#define bacnet_handler_read_property handler_read_property
#define bacnet_Handler_Transmit_Buffer Handler_Transmit_Buffer
#define bacnet_handler_who_is handler_who_is
#define bacnet_Load_Control_State_Machine_Handler Load_Control_State_Machine_Handler
#define bacnet_MAX_MPDU MAX_MPDU
#define bacnet_Notification_Class_find_recipient Notification_Class_find_recipient
#define bacnet_npdu_handler npdu_handler
#define bacnet_OBJECT_ANALOG_INPUT OBJECT_ANALOG_INPUT
#define bacnet_OBJECT_DEVICE OBJECT_DEVICE
#define bacnet_object_functions_t object_functions_t
#define bacnet_PROP_OBJECT_LIST PROP_OBJECT_LIST
#define bacnet_PROP_PRESENT_VALUE PROP_PRESENT_VALUE
#define bacnet_rp_ack_decode_service_request rp_ack_decode_service_request
#define bacnet_rp_ack_print_data rp_ack_print_data
#define bacnet_Send_I_Am Send_I_Am
#define bacnet_Send_Read_Property_Request Send_Read_Property_Request
#define bacnet_Send_WhoIs Send_WhoIs
#define bacnet_trend_log_timer trend_log_timer
#define bacnet_tsm_free_invoke_id tsm_free_invoke_id
#define bacnet_tsm_invoke_id_failed tsm_invoke_id_failed
#define bacnet_tsm_invoke_id_free tsm_invoke_id_free
#define bacnet_tsm_timer_milliseconds tsm_timer_milliseconds


src/bacnet_server.c
/* Modified bacnet_server.c for modbus thread with looping
*  by changing IP address, altering test data, adding linked list functions
*  removing unwanted generic functions and adding modbus functions
*  Sorting out names*/

#include <stdio.h>

#include <libbacnet/address.h>	/*bacnet headers */
#include <libbacnet/device.h>
#include <libbacnet/handlers.h>
#include <libbacnet/datalink.h>
#include <libbacnet/bvlc.h>
#include <libbacnet/client.h>
#include <libbacnet/txbuf.h>
#include <libbacnet/tsm.h>
#include <libbacnet/ai.h>
#include "bacnet_namespace.h"

#include "modbus-tcp.h"		/*modbus headers include all required*/

#define BACNET_DEVICE_ID	44	// my assigned device Id
#define BACNET_INSTANCE_NO	12	// for self-testing
#define BACNET_PORT		0xBAC1	/* default bacnet UDP port */
#define BACNET_INTERFACE	"lo"	/* loopback mode*/
#define BACNET_DATALINK_TYPE	"bvlc"	/* Bacnet virtual link control */
#define BACNET_SELECT_TIMEOUT_MS    0	/* msec */

#define SERVER_PORT		502	// Modbus default port number
//#define SERVER_PORT           0xBAC0          /*  another UDP port*/
#define SERVER_ADDRESS	"127.0.0.1"	// loopback address for testing
//#define SERVER_ADDRESS                "140.159.153.159"    //modbus server address

//#define RUN_AS_BBMD_CLIENT	1		/* true so next part will run*/

#if RUN_AS_BBMD_CLIENT		/* BBMD =>talking to BACnet client */
#define BACNET_BBMD_PORT	0xBAC1
#define BACNET_BBMD_ADDRESS	"140.159.160.7"	/* check to see if BBMD operating */
#define BACNET_BBMD_TTL		90
#endif

/* These are needed for Bacnet maintenance tasks */
static pthread_mutex_t timer_lock = PTHREAD_MUTEX_INITIALIZER;

//---------------------------------------------------------------------
/* SELF TESTING 
 * BACnet client will print "Successful match" whenever it is able to receive
 * this set of data (matching data in EES4100_Testbench/src/random_data /12) */
static uint16_t test_data[] = {
    0xA4EC, 0x6E39, 0x8740, 0x1065, 0x9134, 0xFC8C
};

#define NUM_TEST_DATA (sizeof(test_data)/sizeof(test_data[0]))

//---------------------------------------------------------------------
//Arranging sequence for listing the data

/*1. Define linked list objects*/
typedef struct s_word_object word_object;
struct s_word_object {
    uint16_t word;	
    word_object *next;
};

static word_object *list_head;

/*2. Share list using list lock*/
static pthread_mutex_t list_lock = PTHREAD_MUTEX_INITIALIZER;
static pthread_cond_t list_data_ready = PTHREAD_COND_INITIALIZER;
static pthread_cond_t list_data_flush = PTHREAD_COND_INITIALIZER;

/*3. Add objects to list*/
static void add_to_list(uint16_t word)
{
    word_object *last_object;	//more variables- have been defined?
    if (list_head == NULL) {
	last_object = malloc(sizeof(word_object));
	list_head = last_object;	//if no value, last obj to top?
    } else {
	last_object = list_head;	/*4. Allocate memory to each object */
	while (last_object->next)
	    last_object = last_object->next;
	last_object->next = malloc(sizeof(word_object));
	last_object = last_object->next;
    }
    last_object->word = word;
    last_object->next = NULL;
}

//---------------------------------------------------------------------
//gathering the data from modbus server to send to Bacnet client
//"get list head object"
//modification required here

static int Update_Analog_Input_Read_Property(BACNET_READ_PROPERTY_DATA *
					     rpdata)
{
    static int index;		//gets incremented below

    int instance_no =
	bacnet_Analog_Input_Instance_To_Index(rpdata->object_instance);

    if (rpdata->object_property != bacnet_PROP_PRESENT_VALUE)
	goto not_pv;

    printf("AI_Present_Value request for instance %i\n", instance_no);
    
    /* Without reconfiguring libbacnet, a maximum of 4 values may be sent */
    bacnet_Analog_Input_Present_Value_Set(0, test_data[index++]);
    bacnet_Analog_Input_Present_Value_Set(1, test_data[index++]);
    /* bacnet_Analog_Input_Present_Value_Set(2, test_data[index++]); */

    //setting up the thread for number of instances to be sent?

    if (index == NUM_TEST_DATA)
	index = 0;

  not_pv:			/*If not present value program comes here */
    printf("Not Present_Value\n");	//test print
    return bacnet_Analog_Input_Read_Property(rpdata);
}

//---------------------------------------------------------------------

static bacnet_object_functions_t server_objects[] = {
    {bacnet_OBJECT_DEVICE,
     NULL,
     bacnet_Device_Count,
     bacnet_Device_Index_To_Instance,
     bacnet_Device_Valid_Object_Instance_Number,
     bacnet_Device_Object_Name,
     bacnet_Device_Read_Property_Local,
     bacnet_Device_Write_Property_Local,
     bacnet_Device_Property_Lists,
     bacnet_DeviceGetRRInfo,
     NULL,	/* Iterator */
     NULL,	/* Value_Lists */
     NULL,	/* COV */
     NULL,	/* COV Clear */
     NULL 	/* Intrinsic Reporting */ },
    {bacnet_OBJECT_ANALOG_INPUT,
     bacnet_Analog_Input_Init,
     bacnet_Analog_Input_Count,
     bacnet_Analog_Input_Index_To_Instance,
     bacnet_Analog_Input_Valid_Instance,
     bacnet_Analog_Input_Object_Name,
     Update_Analog_Input_Read_Property,
     bacnet_Analog_Input_Write_Property,
     bacnet_Analog_Input_Property_Lists,
     NULL 	/* ReadRangeInfo */ ,
     NULL 	/* Iterator */ ,
     bacnet_Analog_Input_Encode_Value_List,
     bacnet_Analog_Input_Change_Of_Value,
     bacnet_Analog_Input_Change_Of_Value_Clear,
     bacnet_Analog_Input_Intrinsic_Reporting},
    {MAX_BACNET_OBJECT_TYPE}
};

//---------------------------------------------------------------------

/*Running as BBMD client*/
static void register_with_bbmd(void)
{
#if RUN_AS_BBMD_CLIENT
    /* Thread safety: Shares data with datalink_send_pdu */
    bacnet_bvlc_register_with_bbmd(bacnet_bip_getaddrbyname
				   (BACNET_BBMD_ADDRESS),
				   htons(BACNET_BBMD_PORT),
				   BACNET_BBMD_TTL);
#endif
}

//---------------------------------------------------------------------

/*Setting up minute and second timers*/
static void *minute_tick(void *arg)
{
    while (1) {
	pthread_mutex_lock(&timer_lock);

	/* Expire addresses once the TTL has expired */
	bacnet_address_cache_timer(60);

	/* Re-register with BBMD once BBMD TTL has expired */
	register_with_bbmd();

	/*Update addresses for notification class recipient list 
	 * Requred for INTRINSIC_REPORTING
	 * bacnet_Notification_Class_find_recipient(); */

	/* Sleep for 1 minute */
	pthread_mutex_unlock(&timer_lock);
	sleep(60);
    }
    return arg;
}

static void *second_tick(void *arg)
{
    while (1) {
	pthread_mutex_lock(&timer_lock);

	/* Invalidates stale BBMD foreign device table entries */
	bacnet_bvlc_maintenance_timer(1);

	/* Transaction state machine: Responsible for retransmissions 
	 * and ack checking for confirmed services */
	bacnet_tsm_timer_milliseconds(1000);

	/*removed more bacnet functions not used here */

	/* Sleep for 1 second */
	pthread_mutex_unlock(&timer_lock);
	sleep(1);
    }
    return arg;
}

//---------------------------------------------------------------------
//MODBUS THREAD IN HERE

static void *modbus_side(void *arg)	//allocate and initialize a structure
{
    uint16_t tab_reg[32];
    int rv;			//return value
    int i;			//iteration 
    modbus_t *ctx;

    printf("got to the modbus thread\n");	//test print       

  modbus_restart:		//retry connection at this point

    ctx = modbus_new_tcp(SERVER_ADDRESS, SERVER_PORT);
    //context is modbus server address


    if (ctx == NULL) {
	fprintf(stderr, "Unsuccessful allocation and initialization\n");
	sleep(1);
	goto modbus_restart;
    }
    if (modbus_connect(ctx) == -1) {	//establish connection status
	fprintf(stderr, "Unsuccessful connection to server: %s\n",
		modbus_strerror(errno));
	modbus_free(ctx);
	modbus_close(ctx);
	sleep(1);
	goto modbus_restart;
    } else {
	fprintf(stderr, "Successful connection to server\n");
    }

//read modbus registers 

    while (1) {				//assigned address info
	rv = modbus_read_registers(ctx, BACNET_DEVICE_ID, 2, tab_reg);
	printf("got into the read regs rv value is: %d\n", rv);	//test print
	if (rv == -1) {
	    fprintf(stderr, "Register read failed: %s\n",
		    modbus_strerror(errno));
	    modbus_free(ctx);
	    modbus_close(ctx);
	    sleep(1);
	    goto modbus_restart;
	}
	for (i = 0; i < rv; i++) {
	    //add_to_list(&list_head[i], tab_reg[i]);
	    printf("Register[%d] = [%d] (0x%X)\n", i,
		   tab_reg[i], tab_reg[i]);
	}
	usleep(100000);		//100ms sleep
	//return 0;
    }
    printf("got to the end of modbus thread \n");	//test print       
}

//end of modbus thread
//---------------------------------------------------------------------

#define BN_UNC(service, handler) \
    bacnet_apdu_set_unconfirmed_handler(		\
		    SERVICE_UNCONFIRMED_##service,	\
		    bacnet_handler_##handler)
#define BN_CON(service, handler) \
    bacnet_apdu_set_confirmed_handler(			\
		    SERVICE_CONFIRMED_##service,	\
		    bacnet_handler_##handler)
//---------------------------------------------------------------------
//MAIN 

int main(int argc, char **argv)
{
    uint8_t rx_buf[bacnet_MAX_MPDU];
    uint16_t pdu_len;
    BACNET_ADDRESS src;
    pthread_t minute_tick_id, second_tick_id;
    pthread_t modbus_start_id;	/*added modbus thread */
    bacnet_Device_Set_Object_Instance_Number(BACNET_INSTANCE_NO);
    bacnet_address_init();

    printf("got to the main\n");	//test print

/* Setup device objects */
    bacnet_Device_Init(server_objects);
    BN_UNC(WHO_IS, who_is);
    BN_CON(READ_PROPERTY, read_property);

    bacnet_BIP_Debug = true;
    bacnet_bip_set_port(htons(BACNET_PORT));
    bacnet_datalink_set(BACNET_DATALINK_TYPE);
    bacnet_datalink_init(BACNET_INTERFACE);
    atexit(bacnet_datalink_cleanup);
    memset(&src, 0, sizeof(src));

    register_with_bbmd();

    bacnet_Send_I_Am(bacnet_Handler_Transmit_Buffer);

    pthread_create(&minute_tick_id, 0, minute_tick, NULL);
    pthread_create(&second_tick_id, 0, second_tick, NULL);
    pthread_create(&modbus_start_id, 0, modbus_side, NULL);//make modbus thread 

    /* Start another thread here to retrieve your allocated registers from 
     * the modbus server. This thread should have the following structure 
     * (in a separate function):
     *
     * Initialise:
     *      Connect to the modbus server
     *
     * Loop:
     *      Read the required number of registers from the modbus server
     *      Store the register data into the tail of a linked list 
     */
//---------------------------------------------------------------------
/*holding the lock to guarantee atomicity*/

    while (1) {
	pdu_len =
	    bacnet_datalink_receive(&src, rx_buf, bacnet_MAX_MPDU,
				    BACNET_SELECT_TIMEOUT_MS);
	if (pdu_len) {
	    /* May call any registered handler.
	     * Thread safety: May block, however we still need to guarantee
	     * atomicity with the timers, so hold the lock anyway */
	    pthread_mutex_lock(&timer_lock);
	    bacnet_npdu_handler(&src, rx_buf, pdu_len);
	    pthread_mutex_unlock(&timer_lock);
	}
    }
    return 0;
}
//END OF PROGRAM
